### PR TITLE
rocblas alt impl during backward pass only

### PR DESCRIPTION
**Pull request by jeffdaily**
In preparation of adopting future rocblas library options, it is necessary to track when the backward pass of training is executing.  The scope-based helper class `BackwardPassGuard` is provided to toggle state.

### CODE REVIEW

**Review by albanD — COMMENTED** [review]
Thanks for the PR but I think this would benefit from having at least an issue discussing what we want to achieve with this new TLS and how to make it work.

Some questions we should answer there are:
- Why would a blas gemm kernel need to know if you're in fw or bw? If the idea is that different default precision should be used, then tools like autocast seem to be the right way to go to get such dynamic precision no?
- What do you expect to happen in the context of re-entrant autograd: when your backward is also a forward
- Having ifdef ROCM guards this way is not a great idea in terms of maintenance. We should avoid that as much as possible.
- How would you test something like this? Ensure it does what you expect?

**Comment by albanD** [comment]
Thanks for the details. I'm not sure it addresses them though.
I agree that precision tuning might need to happen. But that was not my question above. The question above was related to forward vs backward behavior that other libraries don't need.
Also related to amp that is already a mechanism to handle this kind of things.

**Comment by atalman** [comment]
@jeffdaily Would it be possible to add small unit test to this PR to test the effect of this flag ? Considering this is proposed as Stable feature for 1.12 release

**Review by albanD — COMMENTED** [review]
From more discussion with @ngimel we think this is ok for now. More details below.


We should add more documentation to this:
- The numerical accuracy page should be updated explaining what this system is doing and when: https://pytorch.org/docs/stable/notes/numerical_accuracy.html In particular, it should have a list of ops that change, the exact hardware and software version when they change.
- Each op that is modified should have a small note linking to the numerical accuracy page mentioning the different precision in backward.

Also note that after the proposed changes, "backward" above will include hooks and final callbacks for more consistency. This should be clearly stated in the doc as well.

As @atalman mentioned above, testing would be needed as well.

File: aten/src/ATen/Context.h [hunk 1]
```diff
@@ -387,4 +387,14 @@ struct TORCH_API NoTF32Guard {
   bool changed = false;
 };
 
+#ifdef USE_ROCM
+struct TORCH_API BackwardPassGuard {
```
[the remark this answers is not in the document]
Comment (jeffdaily) [reply aten/src/ATen/Context.h:391]: Would it be acceptable to keep it named `BackwardPassGuard` without ROCm in the name?  We have downstream pytorch extensions that use this feature already, and it would have a ripple effect.  Nothing that couldn't be resolved, of course.  But it would minimize our developer effort to keep the name as-is.  Let me know.  Thanks.

File: aten/src/ATen/Context.cpp [hunk 2]
```diff
@@ -293,6 +293,22 @@ bool NoTF32Guard::should_disable_tf32() {
   return override_allow_tf32_flag;
 }
 
+#ifdef USE_ROCM
+thread_local bool BackwardPassGuard::is_backward_pass_;
```
[the remark this answers is not in the document]
Comment (jeffdaily) [reply aten/src/ATen/Context.cpp:353]: Resolved by 69d01b06fde25eff612846eb0f72efccc1a36467.

File: torch/csrc/autograd/function.h [hunk 3]
```diff
@@ -151,6 +151,11 @@ struct TORCH_API Node : std::enable_shared_from_this<Node> {
     // probably operate with names.
     at::NoNamesGuard no_names_guard;
 
+#ifdef USE_ROCM
```
[the remark this answers is not in the document]
Comment (jeffdaily) [reply torch/csrc/autograd/function.h:154]: Moved the BackwardPassGuard from torch/csrc/autograd/funtion.h to engine.cpp as suggested.  Confirmed behavior with unit test.

**Comment by jeffdaily** [comment]
> @jeffdaily Would it be possible to add small unit test to this PR to test the effect of this flag ? Considering this is proposed as Stable feature for 1.12 release

Done.  Easiest way was to expose the flag as `torch._C._rocm_is_backward_pass()`.

File: aten/src/ATen/Context.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (albanD) [reply aten/src/ATen/Context.h:407]: Ho that's annoying... If these projects use custom build they will have to be updated to master anyways?

File: torch/csrc/autograd/function.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (albanD) [reply torch/csrc/autograd/function.h:154]: I don't see this in the current PR?

File: torch/csrc/cuda/Module.cpp [hunk 4]
```diff
@@ -630,6 +630,18 @@ PyObject * THCPModule_getCurrentBlasHandle_wrap(PyObject *self, PyObject *noargs
...
 
+PyObject * THCPModule_rocm_is_backward_pass(PyObject *_unused, PyObject *noargs)
+{
+  HANDLE_TH_ERRORS
+#if USE_ROCM
+  bool value = at::BackwardPassGuard::is_backward_pass();
+#else
+  bool value = false;
+#endif
+  return PyBool_FromLong(value);
```
Comment (albanD) [inline torch/csrc/cuda/Module.cpp:641]: nit: you can use `Py_RETURN_TRUE` and `Py_RETURN_FALSE` to avoid the costly conversion:
```cpp
if (at::BackwardPassGuard::is_backward_pass()) {
  Py_RETURN_TRUE;
else {
  Py_RETURN_FALSE;
}

```

**Review by albanD — COMMENTED** [review]
The doc is not here yet right?

File: aten/src/ATen/Context.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (jeffdaily) [reply aten/src/ATen/Context.h]: Renamed `BackwardPassGuard` to `ROCmBackwardPassGuard` in https://github.com/pytorch/pytorch/pull/71881/commits/e4e2c0e8062a29019fe10395b7fca8d8f176145b.

File: torch/csrc/autograd/function.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (jeffdaily) [reply torch/csrc/autograd/function.h]: Pushed with latest batch of commits.

https://github.com/pytorch/pytorch/pull/71881/commits/5d2b8aa811ad711669681f8f4f2f11773a05c87a

File: torch/csrc/cuda/Module.cpp (hunk 4 above)
Comment (jeffdaily) [reply torch/csrc/cuda/Module.cpp]: Fixed in https://github.com/pytorch/pytorch/pull/71881/commits/7e18386a8983f825d26b6b2c425984c4af7a8ae7.

**Comment by jeffdaily** [comment]
@albanD documentation change is underway.  Changes will be part of this PR with modifications to docs/source/notes/numerical_accuracy.rst.

**Review by albanD — COMMENTED** [review]
Thanks for the update.
I think we want a small sentence of the corresponding ops doc as well to make sure it is clear.

Also there were some small CI issues on Friday and the rendered doc wasn't properly made available. A rebase should fix that and allow us to see the rendered version.

**Review by albanD — APPROVED** [review]
SGTM!
